src/lib/js/array.rs

```diff
@@ -251,6 +252,105 @@ pub fn unshift(this: &Value, args: &[Value], _: &mut Interpreter) -> ResultValue
     Ok(to_value(len + arg_c))
 }
 
+/// Array.prototype.indexOf ( searchElement[, fromIndex ] )
+///
+///
+/// indexOf compares searchElement to the elements of the array, in ascending order,
+/// using the Strict Equality Comparison algorithm, and if found at one or more indices,
+/// returns the smallest such index; otherwise, -1 is returned.
+///
+/// The optional second argument fromIndex defaults to 0 (i.e. the whole array is searched).
+/// If it is greater than or equal to the length of the array, -1 is returned,
+/// i.e. the array will not be searched. If it is negative, it is used as the offset
+/// from the end of the array to compute fromIndex. If the computed index is less than 0,
+/// the whole array will be searched.
+/// <https://tc39.es/ecma262/#sec-array.prototype.indexof>
+pub fn index_of(this: &Value, args: &[Value], _: &mut Interpreter) -> ResultValue {
+    // If no arguments, return -1. Not described in spec, but is what chrome does.
+    if args.is_empty() {
+        return Ok(to_value(-1));
+    }
+
+    let search_element = args[0].clone();
+    let len: i32 = from_value(this.get_field_slice("length"))
+        .expect("Expected array property \"length\" is not set.");
+
+    let mut idx = match args.get(1) {
+        Some(from_idx_ptr) => {
+            let from_idx = from_value(from_idx_ptr.clone())
+                .expect("Error parsing \"Array.prototype.indexOf - fromIndex\" argument");
+
+            if from_idx < 0 {
+                len + from_idx
```

Comment:
should this be `len - from_idx` ?

Reply:
It is `len + from_idx` so that negative `from_idx` values are handled correctly, as defined in part 8a of the algroithm defined in the spec [here](https://tc39.es/ecma262/#sec-array.prototype.indexof).

If `from_idx` is negative, we must start the search from the end of the array less the _absolute_ value of `from_idx`. Since we know we need to subtract the _absolute_ value, and we know the value is negative, we can just use addition.

Using a negative `from_idx` is tested around line #496 of this file, they work as expected.

Hope this helps.

Reply:
yep makes sense, thanks!